### PR TITLE
add imagemagick package

### PR DESCRIPTION
**Pull request by Phate6660**
This needs to be tested on bsd and macosx before merging.
It should apparently work the exact same way as installing on Linux according to the docs, but I wanna be sure.
Edit: lol I forgot there are CIs for that.

### CODE REVIEW

File: packages/i/imagemagick/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    on_install("bsd", "linux", "macosx", function(package)
+        local configs = {"--with-utilities=no", "--with-x=no"}
+
+        if package:config("shared") then
+            table.insert(configs, "--enable-shared=yes")
+            table.insert(configs, "--enable-static=no")
+        else
+            table.insert(configs, "--enable-static=yes")
+            table.insert(configs, "--enable-shared=no")
+        end
```
Comment (waruqi) [inline packages/i/imagemagick/xmake.lua:19]: https://github.com/xmake-io/xmake-repo/blob/68fc348e5d650ae1ae4a054b2c0ae926defdd83e/packages/l/libisl/xmake.lua#L14-L15

Comment (waruqi) [reply packages/i/imagemagick/xmake.lua:19]: And please enable pic by default, like this

https://github.com/xmake-io/xmake-repo/blob/68fc348e5d650ae1ae4a054b2c0ae926defdd83e/packages/l/libisl/xmake.lua#L19-L21

File: packages/i/imagemagick/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    set_description("ImageMagick is a FOSS software suite for modifying images. This does NOT provide any of the utilities. It installs the C/C++ Libraries.")
+    add_urls("https://download.imagemagick.org/ImageMagick/download/ImageMagick-$(version).tar.gz")
+    add_versions("7.0.11-13", "6c162d7cbd7b80968a6d083d39eb18d9c9bbd49f500f7f49c9a5dcc0fc36a03b")
+    add_deps("pkg-config")
+    add_deps("ffmpeg", "fftw", "jasper", "libjpeg-turbo", "libpng", "libtiff", "libwebp", "openjpeg", "openexr", "zlib", {optional = true})
+    add_includedirs("include/ImageMagick-7")
+
+    on_install("bsd", "linux", "macosx", function(package)
+        local configs = {"--with-utilities=no", "--with-x=no"}
+
```
Comment (waruqi) [inline packages/i/imagemagick/xmake.lua:12]: remove empty line

File: packages/i/imagemagick/xmake.lua [hunk 3]
```diff
@@ -9,14 +9,8 @@ package("imagemagick")
...
-
-        if package:config("shared") then
-            table.insert(configs, "--enable-shared=yes")
-            table.insert(configs, "--enable-static=no")
-        else
-            table.insert(configs, "--enable-static=yes")
-            table.insert(configs, "--enable-shared=no")
-        end
+         table.insert(configs, "--enable-shared=" .. (package:config("shared") and "yes" or "no")) 
+         table.insert(configs, "--enable-static=" .. (package:config("shared") and "no" or "yes"))
```
Comment (waruqi) [inline packages/i/imagemagick/xmake.lua:13]: Incorrect space indentation

Comment (Phate6660) [reply packages/i/imagemagick/xmake.lua]: Yep fixed that in the pic commit too.

**Comment by waruqi** [comment]
> note: the following packages are unsupported for bsd/x86_64!
  -> pkg-config 

Maybe need add `@bsd` to pkg-config package

**Comment by waruqi** [comment]
> /usr/bin/ld: miff.c:(.text+0x8cbe): undefined reference to `BZ2_bzCompress'
/usr/bin/ld: miff.c:(.text+0x8d3e): undefined reference to `BZ2_bzCompressEnd'
/usr/bin/ld: miff.c:(.text+0x8da9): undefined reference to `lzma_code'
/usr/bin/ld: miff.c:(.text+0x8e29): undefined reference to `lzma_end'
/usr/bin/ld: /home/runner/.config/.xmake/packages/i/imagemagick/7.0.11-13/68813100e8ca4f02803ff9fa5f5d9714/lib/libMagickCore-7.Q16HDRI.a(MagickCore_libMagickCore_7_Q16HDRI_la-msl.o): in function `MSLInternalSubset':
msl.c:(.text+0x5a1): undefined reference to `xmlCreateIntSubset'
/usr/bin/ld: /home/runner/.config/.xmake/packages/i/imagemagick/7.0.11-13/68813100e8ca4f02803ff9fa5f5d9714/lib/libMagickCore-7.Q16HDRI.a(MagickCore_libMagickCore_7_Q16HDRI_la-msl.o): in function `MSLResolveEntity':
msl.c:(.text+0x63c): undefined reference to `xmlLoadExternalEntity'
/usr/bin/ld: /home/runner/.config/.xmake/packages/i/imagemagick/7.0.11-13/68813100e8ca4f02803ff9fa5f5d9714/lib/libMagickCore-7.Q16HDRI.a(MagickCore_libMagickCore_7_Q16HDRI_la-msl.o): in function `MSLGetEntity':
msl.c:(.text+0x6a9): undefined reference to `xmlGetDocEntity'
/usr/bin/ld: /home/runner/.config/.xmake/packages/i/imagemagick/7.0.11-13/68813100e8ca4f02803ff9fa5f5d9714/lib/libMagickCore-7.Q16HDRI.a(MagickCore_libMagickCore_7_Q16HDRI_la-msl.o): in function `MSLGetParameterEntity':
msl.c:(.text+0x70e): undefined reference to `xmlGetParameterEntity'
/usr/bin/ld: /home/runner/.config/.xmake/packages/i/imagemagick/7.0.11-13/68813100e8ca4f02803ff9fa5f5d9714/lib/libMagickCore-7.Q16HDRI.a(MagickCore_libMagickCore_7_Q16HDRI_la-msl.o): in function `MSLEntityDeclaration':
msl.c:(.text+0x7dc): undefined reference to `xmlAddDocEntity'
/usr/bin/ld: msl.c:(.text+0x81a): undefined reference to `xmlAddDtdEntity'
/usr/bin/ld: /home/runner/.config/.xmake/packages/i/imagemagick/7.0.11-13/68813100e8ca4f02803ff9fa5f5d9714/lib/libMagickCore-7.Q16HDRI.a(MagickCore_libMagickCore_7_Q16HDRI_la-msl.o): in function `MSLAttributeDeclaration':
msl.c:(.text+0x8d4): undefined reference to `xmlSplitQName'
/usr/bin/ld: msl.c:(.text+0x923): undefined reference to `xmlAddAttributeDecl'
/usr/bin/ld: msl.c:(.text+0x974): undefined reference to `xmlAddAttributeDecl'
/usr/bin/ld: msl.c:(.text+0x988): undefined reference to `xmlFree'

Maybe we need add some deps or disable them in cmake.install/configs

**Comment by Phate6660** [comment]
Tracking them down now. At first glance I see bzip2 and some xml-related one.

**Comment by Phate6660** [comment]
Hmm. It seems like the some of the Linux CIs are still failing due to openexr.
As for MacOS, this line in particular stands out:
> checkinfo: ...gramdir/core/sandbox/modules/import/core/tool/linker.lua:75: @programdir/core/sandbox/modules/os.lua:262: Undefined symbols for architecture x86_64:

**Comment by waruqi** [comment]
> https://github.com/xmake-io/xmake-repo/pull/424/checks?check_run_id=2616618403#step:4:571

missing freetype, fontconfig, webp , you can add these deps and disable them.

**Comment by Phate6660** [comment]
Huh, libwebp is already a dependency though.
I'll add freetype and fontconfig.

File: packages/i/imagemagick/xmake.lua [hunk 4]
```diff
@@ -3,14 +3,38 @@ package("imagemagick")
...
+                package:add("deps", dep)
+                if name == "tiff" then
+                    package:add("deps", "zlib")
+                end
+            end
+        end
+    end)
+
     on_install("bsd", "linux", "macosx", function(package)
         local configs = {"--with-utilities=no", "--with-x=no", "--without-djvu", "--without-jbig", "--disable-openmp"}
```
Comment (waruqi) [inline packages/i/imagemagick/xmake.lua:39]: if these deps not found or disabled, we should disable them in configs, like this, `--without-xxx`

File: packages/i/imagemagick/xmake.lua [hunk 5]
```diff
@@ -36,7 +43,28 @@ package("imagemagick")
...
+                    elseif name == "exr" then
+                        table.insert(configs, "--with-" .. "openexr")
+                    else
+                        table.insert(configs, "--with-" .. name)
+                    end
+                else
+                    if name == "openjpeg" then
+                        table.insert(configs, "--without-" .. "openjp2")
+                    elseif name == "exr" then
+                        table.insert(configs, "--with-" .. "openexr")
```
Comment (waruqi) [inline packages/i/imagemagick/xmake.lua:61]: with?

Comment (Phate6660) [reply packages/i/imagemagick/xmake.lua:61]: Oh oops, must have missed that.

**Comment by waruqi** [comment]
> Hmm, I can see a threading issue with the Arch CI.
> The solution according to the internet should be simple. Adding `-pthread` to the CFLAGS.

if is_plat("linux") then
    add_syslinks("pthread")
end

File: packages/i/imagemagick/xmake.lua [hunk 6]
```diff
@@ -70,6 +71,11 @@ package("imagemagick")
         if package:config("pic") ~= false then
             table.insert(configs, "--with-pic")
         end
+        if package:config("threads") then
+            package:add("cflags", "-pthread")
```
Comment (waruqi) [inline packages/i/imagemagick/xmake.lua:75]: add_syslinks

Comment (Phate6660) [reply packages/i/imagemagick/xmake.lua:75]: `-pthread` will already add `-lpthread`, and apparently just adding `-lpthread` rarely helps as it only links the lib and not any macros or anything.

I'll change it to add_syslinks though and place it out of the on_install function.

**Comment by Phate6660** [comment]
Need to disable lcms it looks like from the Ubuntu log.

**Comment by Phate6660** [comment]
Oh oof, I think `libjpeg-turbo` needs to `libjpeg`.

**Comment by Phate6660** [comment]
I'm not sure what's causing Arch/Fedora/Ubuntu to fail.
It fails while checking the header `MagickWand/MagickWand.h` for the function `MagickWandGenesis`.
It appears to have a lot of undefined references to... it's own functions that it supplies?

Yet the same thing succeeds in other Linux CIs.

**Comment by waruqi** [comment]
> checking for c links(MagickCore-7.Q16, MagickWand-7.Q16, Magick++-7.Q16, png, z, jpeg, m)

By default, xmake will automatically scan all library files in the installdir/lib directory, and then automatically add links, but their link order may be unreliable.

If there are multiple library files in a package, and there are link order dependencies between them, we can manually call `add_links("xxx", "yyy")` ot `package:add("links", ...` to rewrite them and adjust the link order.

**Comment by Phate6660** [comment]
Ah that would explain it!

MagickWand depends on MagickCore, and Magick++ is separate. So it should be:

`MagickCore -> MagickWand`
`Magick++`

**Comment by Phate6660** [comment]
I'll need to wait a bit before I can test though.
Compiling GCC right now. 
Shouldn't be too long, I've been compiling for about 4-5 hours now.

**Comment by Phate6660** [comment]
OH I need to update the include dirs like I did with the lib dirs.

**Comment by Phate6660** [comment]
It still seems to fail on the same few CIs.

File: packages/i/imagemagick/xmake.lua [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    add_configs("lzma", {description = "Enable LZMA support.", default = false, type = "boolean"})
+    add_configs("openjpeg", {description = "Enable jpeg support through openjpeg.", default = false, type = "boolean"})
+    add_configs("png", {description = "Enable png support.", default = true, type = "boolean"})
+    add_configs("raw", {description = "Enable raw image support.", default = false, type = "boolean"})
+    add_configs("tiff", {description = "Enable tiff support.", default = false, type = "boolean"})
+    add_configs("threads", {description = "Enable threading support.", default = false})
+    add_configs("xml", {description = "Enable XML support.", default = false, type = "boolean"})
+    add_configs("webp", {description = "Enable webp support.", default = false, type = "boolean"})
+    add_includedirs("include/ImageMagick-7/")
+    add_links("MagickCore-7.Q16", "MagickWand-7.Q16", "Magick++-7.Q16")
```
Comment (waruqi) [inline packages/i/imagemagick/xmake.lua:22]: > MagickWand depends on MagickCore, and Magick++ is separate. 

So it should be:

    add_links("MagickWand-7.Q16", "MagickCore-7.Q16", "Magick++-7.Q16")

**Comment by Phate6660** [comment]
(ubuntu-latest, shared) completed properly, but (ubuntu-latest, static) failed. 
Huh. I assume those are the same just with shared vs static libs right?

**Comment by waruqi** [comment]
> (ubuntu-latest, shared) completed properly, but (ubuntu-latest, static) failed.
> Huh. I assume those are the same just with shared vs static libs right?

missing liblqr? maybe we disable it first?

> gemagick/7.0.11-13/fe10404f83fd46caad7cebdf03666a5d/lib/libMagickCore-7.Q16.a(libMagickCore_7_Q16_la-resize.o): in function `LiquidRescaleImage':
resize.c:(.text+0x31d2): undefined reference to `lqr_carver_new_ext'
/usr/bin/ld: resize.c:(.text+0x3252): undefined reference to `lqr_carver_set_preserve_input_image'
/usr/bin/ld: resize.c:(.text+0x3274): undefined reference to `lqr_carver_init'
/usr/bin/ld: resize.c:(.text+0x329a): undefined reference to `lqr_carver_resize'
/usr/bin/ld: resize.c:(.text+0x32ac): undefined reference to `lqr_carver_get_height'
/usr/bin/ld: resize.c:(.text+0x32bb): undefined reference to `lqr_carver_get_width'
/usr/bin/ld: resize.c:(.text+0x336f): undefined reference to `lqr_carver_scan_reset'
/usr/bin/ld: resize.c:(.text+0x34c9): undefined reference to `lqr_carver_scan_ext'

File: packages/i/imagemagick/xmake.lua [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        for name, dep in pairs(configdeps) do
+            if package:config(name) then
+                package:add("deps", dep)
+                if name == "tiff" then
+                    package:add("deps", "zlib")
+                end
+            end
+        end
+        if package:config("threads") and is_plat("linux") then
+            add_syslinks("pthread")
```
Comment (waruqi) [inline packages/i/imagemagick/xmake.lua:48]: package:add("syslinks", "pthread")

add_syslinks is ont the API of description scope, not in script scope.

File: packages/i/imagemagick/xmake.lua [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    add_configs("freetype", {description = "Enable freetype support.", default = false, type = "boolean"})
+    add_configs("jpeg", {description = "Enable jpeg support through libjpeg.", default = true, type = "boolean"})
+    add_configs("lzma", {description = "Enable LZMA support.", default = false, type = "boolean"})
+    add_configs("openjpeg", {description = "Enable jpeg support through openjpeg.", default = false, type = "boolean"})
+    add_configs("png", {description = "Enable png support.", default = true, type = "boolean"})
+    add_configs("raw", {description = "Enable raw image support.", default = false, type = "boolean"})
+    add_configs("tiff", {description = "Enable tiff support.", default = false, type = "boolean"})
+    add_configs("threads", {description = "Enable threading support.", default = false})
+    add_configs("xml", {description = "Enable XML support.", default = false, type = "boolean"})
+    add_configs("webp", {description = "Enable webp support.", default = false, type = "boolean"})
```
Comment (waruqi) [inline packages/i/imagemagick/xmake.lua:20]: please add empty line

File: packages/i/imagemagick/xmake.lua [hunk 10]
```diff
@@ -0,0 +1,97 @@
+package("imagemagick")
+    set_homepage("https://imagemagick.org/script/index.php")
+    set_description("ImageMagick is a FOSS software suite for modifying images. This does NOT provide any of the utilities. It installs the C/C++ Libraries.")
+    add_urls("https://download.imagemagick.org/ImageMagick/download/ImageMagick-$(version).tar.gz")
+    add_versions("7.0.11-13", "6c162d7cbd7b80968a6d083d39eb18d9c9bbd49f500f7f49c9a5dcc0fc36a03b")
+    add_deps("pkg-config")
```
Comment (waruqi) [inline packages/i/imagemagick/xmake.lua:6]: please add empty line and move add_deps("pkg-config") below

```lua
    add_versions(...)

    add_configs(...)
    add_configs(...)


    add_deps("pkg-config")Write<br class="Apple-interchange-newline">
    add_includedirs("include/ImageMagick-7/")    add_links("MagickWand-7.Q16", "MagickCore-7.Q16", "Magick++-7.Q16")
    add_links("MagickWand-7.Q16", "MagickCore-7.Q16", "Magick++-7.Q16")
```

File: packages/i/imagemagick/xmake.lua [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+                            webp       = "libwebp"}
+        for name, dep in pairs(configdeps) do
+            if package:config(name) then
+                package:add("deps", dep)
+                if name == "tiff" then
+                    package:add("deps", "zlib")
+                end
+            end
+        end
+        if package:config("threads") and is_plat("linux") then
```
Comment (waruqi) [inline packages/i/imagemagick/xmake.lua:47]: please use `package:is_plat("linux")`

File: packages/i/imagemagick/xmake.lua (hunk 8 above)
Comment (Phate6660) [reply packages/i/imagemagick/xmake.lua:48]: I can't use package outside of the script scope though?
When I do it results in:
```
error: ./packages/i/imagemagick/xmake.lua:25: attempt to index global 'package' (a function value)
stack traceback:
    [./packages/i/imagemagick/xmake.lua:25]: in main chunk
```
Lines 22-29 for context are:
```lua
    add_deps("pkg-config")
    add_includedirs("include/ImageMagick-7/")
    add_links("MagickWand-7.Q16", "MagickCore-7.Q16", "Magick++-7.Q16")
    if package:config("threads") and package:is_plat("linux") then
        add_syslinks("pthread")
    end

    on_load(function(package)
```

Comment (waruqi) [reply packages/i/imagemagick/xmake.lua:50]: no, please use `package:add("syslinks", "pthread")` if in on_load script scope

**Comment by waruqi** [comment]
Thanks! :+1:

**Comment by Phate6660** [comment]
I'll make sure to update the package once `pkg-config` has bsd support.
Next on my list was fixing the drogon package, then I'll update the `pkg-config` package.

**Comment by waruqi** [comment]
> I'll make sure to update the package once `pkg-config` has bsd support.
> Next on my list was fixing the drogon package, then I'll update the `pkg-config` package.

it's great, thanks!

**Pull request merged**